### PR TITLE
receiver: avoid race of hashring

### PR DESCRIPTION
**Pull request by YaoZengzeng**
Signed-off-by: YaoZengzeng <yaozengzeng@zju.edu.cn>

### CODE REVIEW

**Comment by YaoZengzeng** [comment]
@squat PTAL

File: pkg/receive/handler.go [hunk 1]
```diff
@@ -140,6 +143,9 @@ func (h *Handler) StorageReady() {
 // Hashring sets the hashring for the handler and marks the hashring as ready.
 // If the hashring is nil, then the hashring is marked as not ready.
 func (h *Handler) Hashring(hashring Hashring) {
+	h.mtx.Lock()
+	defer h.mtx.Unlock()
+
 	if hashring == nil {
 		atomic.StoreUint32(&h.hashringReady, 0)
```
Comment (squat) [inline pkg/receive/handler.go]: Now that we have introduced a new syncronization mechanism, we can get rid of the hashringReady field, which we were accessing atomically to determine hashring readiness. Instead, `isReady` can do a RLock, e.g.
```go
h.mtx.RLock()
hr := h.Hashring != nil
h.mtx.RUnlock()
return sr > 0 && hr
```

File: pkg/receive/handler.go [hunk 2]
```diff
@@ -400,14 +417,25 @@ func (h *Handler) replicate(ctx context.Context, tenant string, wreq *prompb.Wri
 	wreqs := make(map[string]*prompb.WriteRequest)
 	replicas := make(map[string]replica)
 	var i uint64
+
+	h.mtx.RLock()
+	// It is possible that hashring is ready in testReady() but become unready now,
+	// so we need to lock here.
+	if h.hashring == nil {
+		h.mtx.RLock()
```
Comment (squat) [inline pkg/receive/handler.go]: This should be `RUnlock`

File: pkg/receive/handler.go [hunk 3]
```diff
@@ -400,14 +417,25 @@ func (h *Handler) replicate(ctx context.Context, tenant string, wreq *prompb.Wri
 	wreqs := make(map[string]*prompb.WriteRequest)
 	replicas := make(map[string]replica)
 	var i uint64
+
+	h.mtx.RLock()
```
Comment (squat) [inline pkg/receive/handler.go]: Just for clarity, plz move this locking below the comment.

File: pkg/receive/handler.go [hunk 4]
```diff
@@ -275,6 +281,15 @@ func (h *Handler) receive(w http.ResponseWriter, r *http.Request) {
 func (h *Handler) forward(ctx context.Context, tenant string, r replica, wreq *prompb.WriteRequest) error {
 	wreqs := make(map[string]*prompb.WriteRequest)
 	replicas := make(map[string]replica)
+
+	h.mtx.RLock()
```
Comment (squat) [inline pkg/receive/handler.go]: Let's move this below the comment.

**Comment by YaoZengzeng** [comment]
@squat Updated.

**Comment by squat** [comment]
cc @brancz for approval and merge

**Comment by brancz** [comment]
Thanks a lot for finding and fixing this @YaoZengzeng!

**Pull request merged**